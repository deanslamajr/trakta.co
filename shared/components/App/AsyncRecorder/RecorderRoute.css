@import "../../../colors.css";

.container {
  text-align: center;
}

.loading-message {
  color: var(--white);
  font-size: 40px;
}