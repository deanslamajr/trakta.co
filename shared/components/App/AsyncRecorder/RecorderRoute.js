import React from 'react';
import Helmet from 'react-helmet';
import { compose } from 'redux';
import { connect } from 'react-redux';
import withStyles from 'isomorphic-style-loader/lib/withStyles';
import Switch from 'react-router-dom/Switch';
import Route from 'react-router-dom/Route';
import Redirect from 'react-router/Redirect';

import config from '../../../../config';

import Recorder from '../../../../client/components/Recorder';
import Staging from './AsyncStaging';
import Cleanup from './AsyncCleanup';

import * as selectors from '../../../reducers';

import styles from './styles.css';


class RecorderRoute extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      isClient: false
    };
  }

  componentDidMount() {
    this.setState({ isClient: true });
  }

  render() {
    const currentTrakName = this.props.trakName || '';

    return (
      <div className={styles.container}>
        <Helmet>
          <title>{`${currentTrakName}::recorder - ${config('appTitle')}`}</title>
        </Helmet>

        {
          this.state.isClient
            ? (
              <Switch>
                <Route 
                  exact 
                  path={this.props.match.url}
                  component={Recorder} />

                { // If a recording does not exist, we shouldn't be able to route to /staging
                  this.props.objectUrl
                    ? (
                      [
                        <Route key={1} path={`${this.props.match.url}/staging`} component={Staging} />,
                        <Route key={2} path={`${this.props.match.url}/cleanup`} component={Cleanup} />
                      ]
                    )
                    : <Redirect to={{ pathname: `${this.props.match.url}` }}/>
                }
              </Switch>
            )
            : (
              <div className={styles.loadingMessage}>
                !!Loading!!
              </div>
            )
        }
        
      </div>
    );
  }
}

function mapStateToProps(state) {
  return {
    objectUrl: selectors.getStagedObjectUrl(state),
    trakName: selectors.getTrakName(state)
  }
}

export { RecorderRoute }

export default compose(
  withStyles(styles),
  connect(mapStateToProps)
)(RecorderRoute);