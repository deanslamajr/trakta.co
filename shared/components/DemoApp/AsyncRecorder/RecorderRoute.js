import React from 'react';
import Helmet from 'react-helmet';
import ReactAudioPlayer from 'react-audio-player';
import axios from 'axios';

import Recorder from './Recorder';
import config from '../../../../config';

class RecorderRoute extends React.Component {
  constructor(props) {
    super(props);

    this.prompts = {
      START: this._renderStartRecordingPrompt.bind(this),
      STOP: this._renderStopRecordingPrompt.bind(this),
      SAVE: this._renderSaveRecordingPrompt.bind(this),
      SAVE_SUCCESS: this._renderSaveSuccessPrompt.bind(this),
      SAVE_ERROR: this._renderSaveErrorPrompt.bind(this),
    };

    this.state = {
      disableRecording: true,
      currentPrompt: this.prompts.START,
      blob: undefined,
    };

    this._startRecording = this._startRecording.bind(this);
    this._stopRecording = this._stopRecording.bind(this);
    this._changePromptToStop = this._changePromptToStop.bind(this);
    this._getSample = this._getSample.bind(this);
    this._saveRecording = this._saveRecording.bind(this);
  }

  _getSample(blob) {
    this.setState({
      currentPrompt: this.prompts.SAVE,
      blob,
    });
  }

  _changePromptToStop() {
    this.setState({ currentPrompt: this.prompts.STOP });
  }

  _startRecording() {
    this.recorder.start(this._changePromptToStop);
  }

  _stopRecording() {
    this.recorder.stop(this._getSample);
  }

  _saveRecording() {
    const data = new FormData();
    data.append('file', new File([this.state.blob], 'sample.mp3'));

    const config = {
      onUploadProgress: (progressEvent) => {
        console.log(
          `upload progress: ${Math.round(progressEvent.loaded * 100 / progressEvent.total)}`,
        );
      },
    };
    axios
      .post('/api/sample', data, config)
      .then(() => this.setState({ currentPrompt: this.prompts.SAVE_SUCCESS }))
      .catch((err) => {
        // @todo log error
        this.setState({ currentPrompt: this.prompts.SAVE_ERROR });
      });
  }

  _renderSaveErrorPrompt() {
    return <div>Ice Cream melted :(</div>;
  }

  _renderSaveSuccessPrompt() {
    return <div>Taco crunchy!</div>;
  }

  _renderSaveRecordingPrompt() {
    this._saveRecording();
    return <div> saving recording... </div>;
  }

  _renderStopRecordingPrompt() {
    return <button onClick={this._stopRecording}>Stop</button>;
  }

  _renderStartRecordingPrompt() {
    return this.state.disableRecording
      ? <div>Waiting on encoder to initialize</div>
      : <button onClick={this._startRecording}>Start</button>;
  }

  _drawBlobs() {
    return (
      <div>
        {this.state.blob
          ? <ReactAudioPlayer src={window.URL.createObjectURL(this.state.blob)} controls />
          : null}
        {this.state.backedupBlob
          ? <ReactAudioPlayer src={window.URL.createObjectURL(this.state.backedupBlob)} controls />
          : null}
      </div>
    );
  }

  componentDidMount() {
    this.body = document.querySelector('body');

    this.recorder = new Recorder({
      sampleRate: 44100,
      bitRate: 128,
      success: () => {
        this.setState({ disableRecording: false });
      },
      error: (msg) => {
        alert(msg);
      },
      fix: (msg) => {
        // @todo this.setState({ currentPrompt: this.prompts.SUPPORT_ERROR })
        console.log(msg);
      },
    });
  }

  render() {
    return (
      <div style={{ textAlign: 'center' }}>
        <Helmet>
          <title>{`${config('appTitle')} - recorder`}</title>
        </Helmet>

        {this.state.currentPrompt()}
        {this._drawBlobs()}
      </div>
    );
  }
}

export default RecorderRoute;
