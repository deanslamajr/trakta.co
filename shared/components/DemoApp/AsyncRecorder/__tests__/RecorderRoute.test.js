/* eslint-disable import/no-extraneous-dependencies */

import React from 'react';
import { shallow } from 'enzyme';

import RecorderRoute from '../RecorderRoute';

describe('<RecorderRoute />', () => {
  test('renders', () => {
    const wrapper = shallow(<RecorderRoute />);
    expect(wrapper).toMatchSnapshot();
  });
});
