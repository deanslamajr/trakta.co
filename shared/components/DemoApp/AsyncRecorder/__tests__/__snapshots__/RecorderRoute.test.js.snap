// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RecorderRoute /> renders 1`] = `
<div
  style={
    Object {
      "textAlign": "center",
    }
  }
>
  <HelmetWrapper
    encodeSpecialCharacters={true}
  >
    <title>
      tracta.co - recorder
    </title>
  </HelmetWrapper>
  <div>
    Waiting on encoder to initialize
  </div>
  <div />
</div>
`;
